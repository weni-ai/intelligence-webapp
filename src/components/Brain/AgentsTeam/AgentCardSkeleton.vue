<template>
  <section class="agent-card-skeleton">
    <UnnnicSkeletonLoading
      tag="div"
      width="30%"
      height="22px"
    />

    <UnnnicSkeletonLoading
      tag="div"
      width="100%"
      height="44px"
    />

    <section class="agent-card-skeleton__skills">
      <UnnnicSkeletonLoading
        tag="div"
        width="100px"
        height="30px"
      />
      <UnnnicSkeletonLoading
        tag="div"
        width="150px"
        height="30px"
      />
    </section>

    <UnnnicSkeletonLoading
      tag="div"
      width="100%"
      height="38px"
    />
  </section>
</template>

<style lang="scss" scoped>
.agent-card-skeleton {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-xs;

  &__skills {
    width: 100%;
    display: flex;
    gap: $unnnic-spacing-nano;
    flex-wrap: wrap;
  }
}
</style>
