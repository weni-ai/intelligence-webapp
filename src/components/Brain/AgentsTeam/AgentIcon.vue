<template>
  <img
    v-if="icon && agentIcon"
    :src="agentIcon"
    class="content__icon"
    data-testid="agent-icon"
  />
</template>

<script setup>
import { computed } from 'vue';
import { agentIcons } from '@/utils/agentIconService';

const props = defineProps({
  icon: {
    type: String,
    default: '',
  },
});

const agentIcon = computed(() => {
  if (!props.icon) return;

  return agentIcons[props.icon] || null;
});
</script>
