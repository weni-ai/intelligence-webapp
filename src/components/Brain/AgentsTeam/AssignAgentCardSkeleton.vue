<template>
  <section
    class="assign-agent-card-skeleton"
    data-testid="assign-agent-card-skeleton"
  >
    <UnnnicSkeletonLoading
      tag="div"
      width="30%"
      height="22px"
      data-testid="skeleton-loading"
    />

    <UnnnicSkeletonLoading
      tag="div"
      width="100%"
      height="44px"
      data-testid="skeleton-loading"
    />

    <section
      class="assign-agent-card-skeleton__skills"
      data-testid="skills-section"
    >
      <UnnnicSkeletonLoading
        tag="div"
        width="100px"
        height="30px"
        data-testid="skills-skeleton-loading"
      />
      <UnnnicSkeletonLoading
        tag="div"
        width="150px"
        height="30px"
        data-testid="skills-skeleton-loading"
      />
    </section>

    <UnnnicSkeletonLoading
      tag="div"
      width="100%"
      height="38px"
      data-testid="skeleton-loading"
    />
  </section>
</template>

<style lang="scss" scoped>
.assign-agent-card-skeleton {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-xs;

  &__skills {
    width: 100%;
    display: flex;
    gap: $unnnic-spacing-nano;
    flex-wrap: wrap;
  }
}
</style>
