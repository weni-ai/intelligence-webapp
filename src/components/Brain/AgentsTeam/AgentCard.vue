<template>
  <section
    data-testid="agent-card"
    :class="['agent-card', { 'agent-card--empty': empty }]"
  >
    <AgentCardSkeleton
      v-if="loading"
      data-testid="agent-card-skeleton"
    />

    <AgentCardEmpty
      v-else-if="empty"
      data-testid="agent-card-empty"
    />

    <section
      v-else
      class="agent-card__content"
    >
      <UnnnicIntelligenceText
        tag="p"
        family="secondary"
        size="body-gt"
        color="neutral-darkest"
        weight="bold"
        data-testid="title"
      >
        {{ title }}
      </UnnnicIntelligenceText>

      <UnnnicIntelligenceText
        v-if="description"
        tag="p"
        family="secondary"
        size="body-gt"
        color="neutral-cloudy"
        data-testid="description"
      >
        {{ description }}
      </UnnnicIntelligenceText>

      <section class="content__skills">
        <Skill
          v-for="skill in skills"
          :key="skill.name"
          :title="skill.name"
          :icon="skill.icon"
          data-testid="skill"
        />
      </section>
    </section>

    <UnnnicButton
      v-if="!loading && !empty && assignment"
      :class="[
        'agent-card__button',
        { 'agent-card__button--assigned': assigned },
      ]"
      :text="
        assigned
          ? $t('router.agents_team.card.assigned')
          : $t('router.agents_team.card.assign')
      "
      :type="assigned ? 'tertiary' : 'primary'"
      :iconLeft="assigned ? 'check' : ''"
      size="small"
      :loading="isAssigning"
      data-testid="assign-button"
      @click="toggleAgentAssignment"
    />
  </section>
</template>

<script setup>
import { ref } from 'vue';

import { useAgentsTeamStore } from '@/store/AgentsTeam';

import AgentCardSkeleton from './AgentCardSkeleton.vue';
import AgentCardEmpty from './AgentCardEmpty.vue';
import Skill from './Skill.vue';

defineEmits(['assign']);

const props = defineProps({
  loading: {
    type: Boolean,
    default: false,
  },
  title: {
    type: String,
    default: '',
  },
  description: {
    type: String,
    default: '',
  },
  /**
   * An array of skills where each skill is an object with the following properties:
   * @property {string} name - The name of the skill
   * @property {string} icon - The icon of the skill
   */
  skills: {
    type: Array,
    default: () => [],
    validator(skills) {
      return skills.every((skill) => 'name' in skill && 'icon' in skill);
    },
  },
  assignment: {
    type: Boolean,
    default: true,
  },
  assigned: {
    type: Boolean,
    default: false,
  },
  uuid: {
    type: String,
    default: '',
  },
  empty: {
    type: Boolean,
    default: false,
  },
});

const isAssigning = ref(false);
const agentsTeamStore = useAgentsTeamStore();

async function toggleAgentAssignment() {
  isAssigning.value = true;

  try {
    await agentsTeamStore.toggleAgentAssignment({
      uuid: props.uuid,
      is_assigned: !props.assigned,
    });
  } catch (error) {
    console.error(error);
  } finally {
    isAssigning.value = false;
  }
}
</script>

<style lang="scss" scoped>
.agent-card {
  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-cleanest;

  padding: $unnnic-spacing-sm;

  display: grid;
  gap: $unnnic-spacing-sm;
  align-content: space-between;

  &--empty {
    display: flex;

    cursor: pointer;
  }

  &__content {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-xs;

    .content__skills {
      display: flex;
      gap: $unnnic-spacing-nano;
      flex-wrap: wrap;
    }
  }

  :deep(.unnnic-button).agent-card__button {
    &--assigned {
      color: $unnnic-color-weni-600;
    }

    [class*='icon'] {
      color: $unnnic-color-weni-600;
    }
  }
}
</style>
