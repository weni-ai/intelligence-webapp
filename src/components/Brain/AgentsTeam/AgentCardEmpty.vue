<template>
  <button
    class="agent-card-empty"
    data-testid="agent-card-empty"
    @click="createNewAgent"
  >
    <UnnnicIcon
      icon="add"
      size="md"
      scheme="neutral-clean"
      data-testid="agent-card-empty-icon"
    />

    <UnnnicIntelligenceText
      tag="p"
      family="secondary"
      size="body-gt"
      color="neutral-dark"
      data-testid="agent-card-empty-title"
    >
      {{ $t('router.agents_team.create_agent') }}
    </UnnnicIntelligenceText>
  </button>
</template>

<script setup>
import { useAgentsTeamStore } from '@/store/AgentsTeam';

const agentsTeamStore = useAgentsTeamStore();
function createNewAgent() {
  window.open(agentsTeamStore.linkToCreateAgent, '_blank');
}
</script>

<style lang="scss" scoped>
.agent-card-empty {
  border: none;
  background: none;

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  height: 100%;
  width: 100%;

  cursor: pointer;
}
</style>
