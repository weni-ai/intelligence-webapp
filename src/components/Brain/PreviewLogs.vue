<template>
  <section
    ref="logList"
    data-testid="preview-logs"
    :class="`preview-logs preview-logs--${props.logsSide}`"
  >
    <p
      v-if="!processedLogs.length"
      class="preview-logs__empty"
      data-testid="preview-logs-empty"
    >
      {{ $t('router.preview.no_logs_registered') }}
    </p>

    <TransitionGroup
      :class="`preview-logs__logs preview-logs__logs--${props.logsSide}`"
      name="logs"
      tag="ol"
      @enter="updateProgressBarHeight('agent')"
    >
      <li
        v-for="(log, logIndex) in processedLogs"
        :key="logIndex"
        :class="`logs__log logs__log--${props.logsSide}`"
        data-testid="preview-logs-log"
      >
        <p
          :class="`log__agent-name log__agent-name--${props.logsSide}`"
          data-testid="preview-logs-log-agent-name"
        >
          {{ log.agent_name }}
        </p>

        <TransitionGroup
          name="steps"
          tag="ol"
          class="log__steps"
          @enter="updateProgressBarHeight('step')"
        >
          <li
            v-for="(step, stepIndex) in log.steps"
            :key="stepIndex"
            :class="`steps__step steps__step--${props.logsSide}`"
            data-testid="preview-logs-log-step"
          >
            <p>{{ step.title }}</p>
            <button
              v-if="step.trace?.trace?.trace"
              class="step__see-full"
              data-testid="preview-logs-log-step-see-full"
              @click="openModalLogFullDetails(step.title, step.trace)"
            >
              {{ $t('router.preview.see_full_details') }}
            </button>
          </li>
        </TransitionGroup>
      </li>
    </TransitionGroup>

    <article
      :class="`preview-logs__progress-bar preview-logs__progress-bar--${props.logsSide}`"
      data-testid="preview-logs-progress-bar"
    />

    <PreviewLogsDetailsModal
      v-model="showDetailsModal"
      :title="selectedLog.summary"
      :trace="selectedLog.trace"
      data-testid="preview-logs-details-modal"
    />
  </section>
</template>

<script setup>
import { ref, onMounted, nextTick, computed, watch } from 'vue';
import { usePreviewStore } from '@/store/Preview';
import { useAgentsTeamStore } from '@/store/AgentsTeam';
import PreviewLogsDetailsModal from './Preview/PreviewLogsDetailsModal.vue';

const emit = defineEmits(['scroll-to-bottom']);

const props = defineProps({
  logs: {
    type: Array,
    required: true,
  },
  logsSide: {
    type: String,
    default: 'left',
    validator(value) {
      return ['left', 'right'].includes(value);
    },
  },
  agents: {
    type: Object,
    default: () => ({
      agents: [],
      manager: {},
    }),
  },
});

const previewStore = usePreviewStore();
const agentsTeamStore = useAgentsTeamStore();

const showDetailsModal = ref(false);
const selectedLog = ref({
  summary: '',
  trace: '',
});

const processedLogs = computed(() => {
  const allAgents =
    Object.keys(props.agents.manager).length && props.agents.agents.length
      ? props.agents
      : agentsTeamStore.allAgents;
  if (!allAgents) return [];

  const traces = props.logs;
  const { agents, manager } = allAgents;

  return traces.reduce((logsByAgent, trace) => {
    const { agentName = '' } = trace.config || {};

    const agent = agents.find((agent) => agent.id === agentName) || manager;

    if (!agent) return logsByAgent;

    const lastLog = logsByAgent.at(-1);
    if (lastLog?.id !== agent.id) {
      logsByAgent.push({
        id: agent.id,
        agent_name: agent.name || 'Manager',
        steps: [],
      });
    }

    logsByAgent.at(-1)?.steps.push({
      title: getTraceSummary(trace) || 'Unknown',
      trace,
    });
    return logsByAgent;
  }, []);
});

function getTraceSummary(trace) {
  if (trace.config?.summary) {
    return trace.config.summary;
  }

  function capitalizeWord(word) {
    return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
  }

  function formatTraceKey(key) {
    return key
      .split(/(?=[A-Z])|_/)
      .map(capitalizeWord)
      .join(' ');
  }

  function findTraceKey(traceObject) {
    return Object.keys(traceObject).find((key) =>
      key.toLowerCase().includes('trace'),
    );
  }

  if (!trace.trace || typeof trace.trace !== 'object') {
    return 'Unknown';
  }

  const traceKey = findTraceKey(trace.trace.trace || trace.trace);
  return traceKey ? formatTraceKey(traceKey) : 'Unknown';
}

const progressHeight = ref(0);

onMounted(() => {
  nextTick(() => {
    updateProgressBarHeight('mount');
  });
});

function openModalLogFullDetails(summary, trace) {
  selectedLog.value = { summary, trace };
  showDetailsModal.value = true;
}

const logTranslateY = 24;

function updateProgressBarHeight(type = 'agent') {
  nextTick(() => {
    if (!['mount', 'agent', 'step'].includes(type)) {
      throw new Error('Invalid type passed to updateProgressHeight function');
    }

    const logList = document.querySelector('.preview-logs');
    if (!logList) return;

    const firstLog = logList.querySelector('.log__agent-name:first-child');
    const lastLog = logList.querySelector(
      '.logs__log:last-child .steps__step:last-child p',
    );

    if (firstLog && lastLog) {
      const firstLogRect = firstLog.getBoundingClientRect();
      const lastLogRect = lastLog.getBoundingClientRect();
      const animationOffset = type === 'agent' ? logTranslateY : 0;
      progressHeight.value =
        animationOffset +
        lastLogRect.bottom -
        firstLogRect.top -
        firstLogRect.height / 2 -
        lastLogRect.height / 2;
    }

    emit('scroll-to-bottom');
  });
}

watch(
  () => previewStore.collaboratorsTraces,
  (newTraces) => {
    if (newTraces.length === 0) {
      progressHeight.value = 0;
    }
  },
);
</script>

<style scoped lang="scss">
.logs-enter-active {
  transition: all 0.5s ease-out;
}
.logs-enter-from {
  $logTranslateY: -24px;
  opacity: 0;
  transform: translateY($logTranslateY);
}

.steps-enter-active {
  transition: all 0.5s ease-out;
}
.steps-enter-from {
  opacity: 0;
}

.preview-logs {
  position: relative;

  &--left {
    margin: 0 auto;
  }

  .preview-logs__empty {
    margin: 0;
    padding: 0;

    color: $unnnic-color-neutral-clean;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  .preview-logs__logs {
    position: relative;

    margin: 0;

    padding: 0;

    list-style: none;

    &--left {
      margin-left: $unnnic-spacing-sm;
    }

    &--right {
      text-align: end;
      margin-right: $unnnic-spacing-sm;
    }

    .logs__log {
      $progressDotOffset: -($unnnic-spacing-sm + $unnnic-spacing-nano) + 0.5;

      margin-bottom: $unnnic-spacing-sm;

      %progressDot {
        &::before {
          content: 'â€¢';

          position: absolute;
          z-index: 1;

          color: $unnnic-color-neutral-cleanest;
        }
      }

      .log__agent-name {
        margin-bottom: $unnnic-spacing-xs;

        color: $unnnic-color-neutral-darkest;
        font-family: $unnnic-font-family-secondary;
        font-weight: $unnnic-font-weight-bold;
        font-size: $unnnic-font-size-body-gt;
        line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;

        @extend %progressDot;

        &--left {
          &::before {
            left: $progressDotOffset;
          }
        }

        &--right {
          &::before {
            right: $progressDotOffset;
          }
        }
      }

      .log__steps {
        margin: 0;
        margin-left: $unnnic-spacing-md - $unnnic-spacing-ant;

        padding: 0;

        display: grid;
        gap: $unnnic-spacing-xs;

        list-style: none;

        .steps__step {
          &:first-letter {
            text-transform: uppercase;
          }
          color: $unnnic-color-neutral-cloudy;
          font-family: $unnnic-font-family-secondary;
          font-size: $unnnic-font-size-body-gt;
          line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;

          @extend %progressDot;

          &--left {
            &::before {
              left: $progressDotOffset;
            }
          }

          &--right {
            &::before {
              right: $progressDotOffset;
            }
          }

          .step__see-full {
            padding: 0;
            border: none;
            background-color: transparent;

            color: $unnnic-color-neutral-clean;

            font-family: $unnnic-font-family-secondary;
            font-size: $unnnic-font-size-body-md;
            line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;

            cursor: pointer;
          }
        }
      }

      &:last-child {
        .steps__step:last-of-type {
          color: $unnnic-color-weni-600;

          &::before {
            color: $unnnic-color-weni-600;
          }

          .step__see-full {
            color: $unnnic-color-weni-600;
          }
        }
      }
    }
  }

  .preview-logs__progress-bar {
    position: absolute;
    top: calc(($unnnic-font-size-body-gt + $unnnic-line-height-md) / 2);

    width: $unnnic-border-width-thinner;
    height: v-bind('progressHeight + "px"');

    background-color: $unnnic-color-neutral-cleanest;

    transition: height 0.6s ease;

    &--right {
      right: 0;
    }
  }
}
</style>
