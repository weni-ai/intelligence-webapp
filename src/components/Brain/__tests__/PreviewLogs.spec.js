import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { createTestingPinia } from '@pinia/testing';
import { nextTick } from 'vue';

import PreviewLogs from '../PreviewLogs.vue';

import { usePreviewStore } from '@/store/Preview';
import { useAgentsTeamStore } from '@/store/AgentsTeam';

const mockAllAgents = {
  agents: [
    { id: 'agent-1', name: 'Test Agent 1' },
    { id: 'agent-2', name: 'Test Agent 2' },
  ],
  manager: { id: 'manager', name: 'Manager' },
};

const mockLogs = [
  {
    type: 'trace_update',
    data: {
      trace: {},
    },
    config: {
      agentName: 'agent-1',
      icon: 'icon-1',
      summary: 'First trace summary',
    },
  },
  {
    type: 'trace_update',
    data: null,
    config: {
      agentName: 'agent-1',
      icon: 'icon-2',
      summary: 'Second trace summary',
    },
  },
  {
    type: 'trace_update',
    data: null,
    config: {
      agentName: 'manager',
      summary: 'Manager trace',
    },
  },
];

const createWrapper = (props = {}) => {
  const pinia = createTestingPinia({
    initialState: {
      preview: {
        logs: [...mockLogs],
      },
    },
  });

  return mount(PreviewLogs, {
    props: {
      logs: mockLogs,
      logsSide: 'left',
      ...props,
    },
    global: {
      plugins: [pinia],
      stubs: {
        TransitionGroup: false,
        PreviewLogsDetailsModal: true,
      },
    },
    attachTo: document.body,
  });
};

describe('PreviewLogs.vue', () => {
  let wrapper;
  let previewStore;
  let agentsTeamStore;

  const previewLogs = () => wrapper.find('[data-testid="preview-logs"]');
  const emptyMessage = () => wrapper.find('[data-testid="preview-logs-empty"]');
  const logItems = () => wrapper.findAll('[data-testid="preview-logs-log"]');
  const logAgentNames = () =>
    wrapper.findAll('[data-testid="preview-logs-log-agent-name"]');
  const logStepIcons = () =>
    wrapper.findAll('[data-testid="preview-logs-log-step-icon"]');
  const logStepIconsIcon = () =>
    wrapper.findAll('[data-testid="preview-logs-log-step-icon-icon"]');
  const stepItems = () =>
    wrapper.findAll('[data-testid="preview-logs-log-step"]');
  const seeFullButtons = () =>
    wrapper.findAll('[data-testid="preview-logs-log-step-see-full"]');
  const progressBar = () =>
    wrapper.find('[data-testid="preview-logs-progress-bar"]');
  const modal = () =>
    wrapper.findComponent('[data-testid="preview-logs-details-modal"]');

  beforeEach(() => {
    Element.prototype.getBoundingClientRect = vi.fn(() => ({
      top: 0,
      bottom: 100,
      height: 20,
    }));

    wrapper = createWrapper();
    previewStore = usePreviewStore();
    agentsTeamStore = useAgentsTeamStore();
    agentsTeamStore.allAgents = mockAllAgents;
  });

  afterEach(() => {
    wrapper.unmount();
    vi.clearAllMocks();
  });

  describe('Component rendering', () => {
    it('matches snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders correctly with required props', () => {
      expect(previewLogs().exists()).toBe(true);
    });

    it('renders the correct number of logs and steps', () => {
      expect(logItems().length).toBe(2);
      expect(stepItems().length).toBe(3);
    });

    it('renders the correct icon for each step', () => {
      expect(logStepIcons().length).toBe(2);
      expect(logStepIconsIcon().at(0).text()).toBe('icon-1');
      expect(logStepIconsIcon().at(1).text()).toBe('icon-2');
    });

    it('renders the correct agent names', () => {
      expect(logAgentNames().at(0).text()).toBe('Test Agent 1');
      expect(logAgentNames().at(1).text()).toBe('Manager');
    });

    it('displays empty message when no logs are available', async () => {
      await wrapper.setProps({ logs: [] });

      expect(emptyMessage().exists()).toBe(true);
      expect(emptyMessage().text()).toBe(
        wrapper.vm.$t('router.preview.no_logs_registered'),
      );
    });
  });

  describe('Props validation', () => {
    it('accepts valid logsSide values', () => {
      const leftWrapper = createWrapper({ logsSide: 'left' });
      const rightWrapper = createWrapper({ logsSide: 'right' });

      expect(leftWrapper.props('logsSide')).toBe('left');
      expect(rightWrapper.props('logsSide')).toBe('right');
    });

    it('applies correct CSS classes based on logsSide prop', async () => {
      expect(previewLogs().classes()).toContain('preview-logs--left');

      await wrapper.setProps({ logsSide: 'right' });
      expect(previewLogs().classes()).toContain('preview-logs--right');
    });

    it('uses provided agents prop when available', async () => {
      const customAgents = {
        agents: [{ id: 'custom-agent', name: 'Custom Agent' }],
        manager: { id: 'custom-manager', name: 'Custom Manager' },
      };

      const customLogs = [
        {
          type: 'trace_update',
          data: null,
          config: {
            agentName: 'custom-agent',
            summary: 'Custom agent trace',
          },
        },
      ];

      await wrapper.setProps({
        agents: customAgents,
        logs: customLogs,
      });

      expect(logAgentNames().at(0).text()).toBe('Custom Agent');
    });
  });

  describe('Logs processing', () => {
    it('correctly processes logs from traces', () => {
      const processedLogs = wrapper.vm.processedLogs;

      expect(processedLogs.length).toBe(2);
      expect(processedLogs[0].agent).toBe('Test Agent 1');
      expect(processedLogs[0].steps.length).toBe(2);
      expect(processedLogs[1].agent).toBe('Manager');
      expect(processedLogs[1].steps.length).toBe(1);
    });

    it('handles empty traces correctly', async () => {
      await wrapper.setProps({ logs: [] });

      expect(wrapper.vm.processedLogs).toEqual([]);
    });

    it('handles missing agent correctly', async () => {
      const logsWithUnknownAgent = [
        {
          type: 'trace_update',
          summary: 'Unknown agent trace',
          data: null,
          config: {
            agentId: 'unknown-agent',
          },
        },
      ];

      await wrapper.setProps({ logs: logsWithUnknownAgent });

      expect(wrapper.vm.processedLogs.length).toBe(1);
      expect(wrapper.vm.processedLogs[0].agent).toBe('Manager');
    });

    it('groups consecutive logs from the same agent', async () => {
      const consecutiveLogs = [
        {
          type: 'trace_update',
          data: null,
          config: {
            agentName: 'agent-1',
            summary: 'First step',
          },
        },
        {
          type: 'trace_update',
          data: null,
          config: {
            agentName: 'agent-1',
            summary: 'Second step',
          },
        },
        {
          type: 'trace_update',
          data: null,
          config: {
            agentName: 'agent-2',
            summary: 'Third step',
          },
        },
      ];

      await wrapper.setProps({ logs: consecutiveLogs });
      const processedLogs = wrapper.vm.processedLogs;

      expect(processedLogs.length).toBe(2);
      expect(processedLogs[0].steps.length).toBe(2);
      expect(processedLogs[1].steps.length).toBe(1);
    });

    it('handles logs without config correctly', async () => {
      const logsWithoutConfig = [
        {
          type: 'trace_update',
          data: { trace: { someTraceData: 'value' } },
        },
      ];

      await wrapper.setProps({ logs: logsWithoutConfig });
      const processedLogs = wrapper.vm.processedLogs;

      expect(processedLogs.length).toBe(1);
      expect(processedLogs[0].agent).toBe('Manager');
    });
  });

  describe('getLogSummary function', () => {
    it('returns config summary when available', () => {
      const logWithSummary = {
        config: { summary: 'Custom Summary' },
      };

      const summary = wrapper.vm.getLogSummary(logWithSummary);
      expect(summary).toBe('Custom Summary');
    });

    it('generates camelCase summary from trace keys', () => {
      const logWithTrace = {
        data: {
          trace: {
            someComplexTraceKey: 'some data',
            otherData: 'value',
          },
        },
      };

      const summary = wrapper.vm.getLogSummary(logWithTrace);
      expect(summary).toBe('Some Complex Trace Key');
    });

    const unknownCases = [
      {
        description: 'returns "Unknown" when no trace key is found',
        log: { data: { someData: 'value' } },
      },
      {
        description: 'returns "Unknown" when data is null',
        log: { data: null },
      },
      {
        description: 'returns "Unknown" when data is not an object',
        log: { data: 'string data' },
      },
      {
        description: 'returns "Unknown" when log has no data',
        log: { type: 'some_type' },
      },
    ];

    unknownCases.forEach(({ description, log }) => {
      it(description, () => {
        const summary = wrapper.vm.getLogSummary(log);
        expect(summary).toBe('Unknown');
      });
    });
  });

  describe('Progress bar functionality', () => {
    it('calculates progress bar height on mount', async () => {
      await nextTick();
      expect(typeof wrapper.vm.progressHeight).toBe('number');
      expect(progressBar().exists()).toBe(true);
    });

    it('updates progress bar height with different types', async () => {
      const spy = vi.spyOn(wrapper.vm, 'updateProgressBarHeight');

      wrapper.vm.updateProgressBarHeight('agent');
      wrapper.vm.updateProgressBarHeight('step');
      wrapper.vm.updateProgressBarHeight('mount');

      expect(spy).toHaveBeenCalledWith('agent');
      expect(spy).toHaveBeenCalledWith('step');
      expect(spy).toHaveBeenCalledWith('mount');
    });

    it('applies correct progress bar positioning for right side', async () => {
      await wrapper.setProps({ logsSide: 'right' });

      expect(progressBar().classes()).toContain(
        'preview-logs__progress-bar--right',
      );
    });
  });

  describe('Modal functionality', () => {
    it('opens modal with correct details when "see full details" is clicked', async () => {
      expect(modal().props('modelValue')).toBeFalsy();

      await seeFullButtons().at(0).trigger('click');

      expect(modal().props('modelValue')).toBeTruthy();

      expect(modal().props('title')).toBe('First trace summary');
      expect(modal().props('log')).toEqual(mockLogs[0]);
    });

    it('closes modal correctly', async () => {
      await seeFullButtons().at(0).trigger('click');
      expect(modal().props('modelValue')).toBeTruthy();

      await modal().vm.$emit('update:modelValue', false);

      expect(wrapper.vm.showDetailsModal).toBe(false);
    });
  });

  describe('Watchers', () => {
    it('resets progress height when logs become empty', async () => {
      wrapper.vm.progressHeight = 100;

      await wrapper.setProps({ logs: [] });
      await nextTick();

      expect(wrapper.vm.progressHeight).toBe(0);
    });

    it('does not reset progress height when logs are not empty', async () => {
      wrapper.vm.progressHeight = 100;

      await wrapper.setProps({ logs: [mockLogs[0]] });
      await nextTick();

      expect(wrapper.vm.progressHeight).toBe(100);
    });
  });

  describe('Events', () => {
    it('emits scroll-to-bottom when updateProgressBarHeight is called', async () => {
      vi.useFakeTimers();
      await wrapper.vm.updateProgressBarHeight();
      await nextTick();

      vi.runAllTimers();
      vi.useRealTimers();

      expect(wrapper.emitted('scroll-to-bottom')).toBeTruthy();
    });

    it('emits scroll-to-bottom for different update types', async () => {
      vi.useFakeTimers();

      wrapper.vm.updateProgressBarHeight('agent');
      wrapper.vm.updateProgressBarHeight('step');

      vi.runAllTimers();
      vi.useRealTimers();

      expect(wrapper.emitted('scroll-to-bottom')).toHaveLength(2);
    });
  });

  describe('Step rendering', () => {
    it('renders steps without icons when icon is not provided', async () => {
      const logsWithoutIcons = [
        {
          data: null,
          config: {
            agentName: 'agent-1',
            summary: 'Step without icon',
          },
        },
        {
          data: null,
          config: {
            agentName: 'agent-1',
            summary: 'Second step without icon',
          },
        },
      ];

      await wrapper.setProps({ logs: logsWithoutIcons });

      expect(logStepIcons().length).toBe(0);
    });

    it('shows see full button only when log data exists', async () => {
      const logsWithAndWithoutData = [
        {
          data: { trace: { someTrace: 'data' } },
          config: {
            agentName: 'agent-1',
            summary: 'Step with data',
          },
        },
        {
          data: null,
          config: {
            agentName: 'agent-1',
            summary: 'Step without data',
          },
        },
      ];

      await wrapper.setProps({ logs: logsWithAndWithoutData });

      expect(seeFullButtons().length).toBe(1);
    });
  });

  describe('Edge cases', () => {
    it('handles logs with missing config.agentName', async () => {
      const logsWithMissingAgentName = [
        {
          data: null,
          config: {
            summary: 'Log without agent name',
          },
        },
      ];

      await wrapper.setProps({ logs: logsWithMissingAgentName });

      expect(wrapper.vm.processedLogs.length).toBe(1);
      expect(wrapper.vm.processedLogs[0].agent).toBe('Manager');
    });

    it('handles agent not found in agents list', async () => {
      const logsWithNonExistentAgent = [
        {
          data: null,
          config: {
            agentName: 'non-existent-agent',
            summary: 'Log from non-existent agent',
          },
        },
      ];

      await wrapper.setProps({ logs: logsWithNonExistentAgent });

      expect(wrapper.vm.processedLogs.length).toBe(1);
      expect(wrapper.vm.processedLogs[0].agent).toBe('Manager');
    });
  });
});
