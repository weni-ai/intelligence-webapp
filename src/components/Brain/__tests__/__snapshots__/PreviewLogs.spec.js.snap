// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PreviewLogs.vue > Component rendering > matches snapshot 1`] = `
"<section data-v-6e9a8500="" data-testid="preview-logs" class="preview-logs preview-logs--left">
  <!--v-if-->
  <ol data-v-6e9a8500="" class="preview-logs__logs preview-logs__logs--left">
    <li data-v-6e9a8500="" class="logs__log logs__log--left" data-testid="preview-logs-log">
      <p data-v-6e9a8500="" class="log__agent-name log__agent-name--left" data-testid="preview-logs-log-agent-name">Test Agent 1</p>
      <ol data-v-6e9a8500="" class="log__steps">
        <li data-v-6e9a8500="" class="steps__step steps__step--left" data-testid="preview-logs-log-step">
          <p data-v-6e9a8500="">First trace summary</p><button data-v-6e9a8500="" class="step__see-full" data-testid="preview-logs-log-step-see-full">See full details</button>
        </li>
        <li data-v-6e9a8500="" class="steps__step steps__step--left" data-testid="preview-logs-log-step">
          <p data-v-6e9a8500="">Second trace summary</p><button data-v-6e9a8500="" class="step__see-full" data-testid="preview-logs-log-step-see-full">See full details</button>
        </li>
      </ol>
    </li>
    <li data-v-6e9a8500="" class="logs__log logs__log--left" data-testid="preview-logs-log">
      <p data-v-6e9a8500="" class="log__agent-name log__agent-name--left" data-testid="preview-logs-log-agent-name">Manager</p>
      <ol data-v-6e9a8500="" class="log__steps">
        <li data-v-6e9a8500="" class="steps__step steps__step--left" data-testid="preview-logs-log-step">
          <p data-v-6e9a8500="">Manager trace</p><button data-v-6e9a8500="" class="step__see-full" data-testid="preview-logs-log-step-see-full">See full details</button>
        </li>
      </ol>
    </li>
  </ol>
  <article data-v-6e9a8500="" class="preview-logs__progress-bar preview-logs__progress-bar--left" data-testid="preview-logs-progress-bar"></article>
  <preview-logs-details-modal-stub data-v-6e9a8500="" title="" trace="" modelvalue="false" data-testid="preview-logs-details-modal"></preview-logs-details-modal-stub>
</section>"
`;
